src/mjolnir/pbfgraphparser.cc

```diff
@@ -5046,31 +5063,33 @@ OSMData PBFGraphParser::ParseWays(const boost::property_tree::ptree& pt,
 
   // Create OSM data. Set the member pointer so that the parsing callback methods can use it.
   OSMData osmdata{};
-  graph_callback callback(pt, osmdata);
+  graph_parser parser(pt, osmdata);
 
   LOG_INFO("Parsing files for ways: " + boost::algorithm::join(input_files, ", "));
 
-  callback.reset(new sequence<OSMWay>(ways_file, true),
-                 new sequence<OSMWayNode>(way_nodes_file, true),
-                 new sequence<OSMAccess>(access_file, true), nullptr, nullptr, nullptr, nullptr);
+  parser.reset(new sequence<OSMWay>(ways_file, true), new sequence<OSMWayNode>(way_nodes_file, true),
+               new sequence<OSMAccess>(access_file, true), nullptr, nullptr, nullptr, nullptr);
   // Parse the ways and find all node Ids needed (those that are part of a
   // way's node list. Iterate through each pbf input file.
   LOG_INFO("Parsing ways...");
   for (auto& file : input_files) {
-    callback.current_way_node_index_ = callback.last_node_ = callback.last_way_ =
-        callback.last_relation_ = 0;
-    OSMPBF::parse(file,
-                  static_cast<OSMPBF::Interest>(OSMPBF::Interest::WAYS |
-                                                OSMPBF::Interest::CHANGESETS),
-                  callback);
+    parser.current_way_node_index_ = parser.last_node_ = parser.last_way_ = parser.last_relation_ = 0;
+
+    osmium::io::Reader reader(file, osmium::osm_entity_bits::way);
+    while (const osmium::memory::Buffer buffer = reader.read()) {
+      for (const osmium::memory::Item& item : buffer) {
+        parser.way(static_cast<const osmium::Way&>(item));
+      }
+    }
+    reader.close(); // Explicit close to get an exception in case of an error.
   }
```

Comment:
As an idea for pbf parsing speed up, we can extract Lua invocation (`LuaTagTransform::Transform()`) from `parser.way()` and perform it on a thread pool - no need to sync access to osmdata and `osmium::memory::Buffer` is an obvious way how to batch such load.